Add a dockerfile to handle dependencies.

annabunches opened the pull request:
While discussing this application, I realized it hadn't been dockerized yet, and that capturing a working set of all its dependencies was really valuable, even if we don't fully dockerize it for production use.

I'll try to get a docker-compose with a full stack working in a future PR.

annabunches committed:
Add a dockerfile to handle dependencies.

annabunches committed:
Don't wipe out entire /etc/environment.

michaelhiiva commented on README.md:
```diff
@@ -0,0 +1,27 @@
+# AGA Ratings Program
+
+This program scans the AGA Database for tournaments that have been added but not yet rated.
+
+It is intended to be run after the tournaments have been uploaded into the database.
```
I find this description using "this" and "it" a little off because we have many applications that are involved in getting our tournaments rated and added to our database. For example: are we running application with`$ bayrate`, `$ bayrate --commit` or `$ ./check`? Also, we have several binaries created:
```
aga@8939089dead4:/srv$ ls
basicrate  bayrate  check  copyMembership  syncDBs
```
by the Makefile. Some of which I am not sure exactly what they do. @vash3g Do you have information on these applications?

vash3g replied:
copyMembership and SyncDBs are legacy and no longer needed. check and bayrate are part of the process in rating tournaments. check will go through and look for tournaments which have a "0" for the rated column in the games table and mark those for ratings. bayrate will perform the actual ratings for the application, first as a check. 'bayrate --commit' will actually rate the tournament and write to the database.

annabunches replied:
Ok, cool, my current understanding, then, is that the usual workflow for this is something like the following. Correct me if I'm wrong please, otherwise I'll update the documentation to match. :)

1. Receive tournament data
2. <Run scripts outside of this repository, correcting errors>
3. Run `./check`
4. Run `./bayrate`
5. Run `./bayrate --commit`

michaelhiiva replied:
@vash3g we should consider removing "copyMembership" and "SyncDBs" from the AGA-Ratings-Program if they are no longer needed. (see: #9)

michaelhiiva replied:
@annabunches That sounds about right. Though, additionally on 2, I think we are committing an unrated tournament to the AGAGD before we run `./check`.

vash3g replied:
Added update to the readme file to include items listed in ToDo.

Adding PR to address issue #9 with updated makefile.

michaelhiiva commented on README.md:
````diff
@@ -0,0 +1,27 @@
+docker run --rm -it agaratings <binary-to-execute>
+```
+
+e.g., to check for tournaments that haven't been rated yet, `docker run --rm -it agaratings check`
````
I saw the ToDo on line 9. Is there a plan to resolve the ToDo on line 9 in this pull request or a later pull request?

vash3g replied:
I think part of the ToDo is the documents ive created in a shared google folder to explain how to run the ratings application after it has been configured

annabunches replied:
@vash3g Yeah I think it's just also useful to have it documented here with the code. Much easier for volunteers to understand the code's basic workflow that way. :)

@EarthSchlange I was going to do it in a future PR, but with the above comments, now I'm going to do it in this PR.

vash3g replied:
Added updated readme that should resolve this.

michaelhiiva commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM ubuntu:18.04
+
+WORKDIR /srv
+RUN useradd -ms /bin/bash aga
+ENV PATH="/srv:${PATH}"
```
I think we should reorder the RUN and ENV. We also don't need to equal sign as we only have one variable here.

annabunches replied:
Oh, that ENV doesn't even need to be here any more, should be removed. Will fix. (see the modification of /etc/environment below, which doesn't clobber the path with the builder's path)

vash3g replied:
When this is updated ill merge the PR.

annabunches replied:
Turns out that `ENV` was load-bearing after all. Guess I'll just have to live with it. Update coming soon.

Commit:
Update README.md

Updated readme to reflect apps created in makefile, basic instructions for use,

Commit:
Update README.md

removed basicrate

vash3g commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+ENV PATH="/srv:${PATH}"
+RUN echo "export PATH=/srv:$PATH" > /etc/environment
+RUN apt update && apt install -y libgsl23 libgslcblas0 libmysql++3v5 libboost-date-time1.65.1 libglib2.0-0
+COPY --from=build --chown=aga:aga /build/bayrate /build/check /build/basicrate /build/copyMembership /build/syncDBs /srv/
```
Submitted a new update to not build basicrate, copyMembership, and syncDBs. Do these just need to be removed from this line?

michaelhiiva replied:
Yes. We should probably remove these files too. That is the memUpdate.cpp which is getting built and syncDBs.cpp.

vash3g replied:
removed lines in updated file

Commit:
Merge remote-tracking branch 'upstream/master' into pr/8

Commit:
Added to PR8

Removed old cpp files, updated dockerfile and makefile to remove docs.

annabunches committed:
Revert ENV-vs-/etc/environment change, and modify apt commands for best practices.

michaelhiiva commented on Makefile:
```diff
@@ -10,7 +10,7 @@ LIBS=-lm -lgsl -lgslcblas -lmysqlpp -lboost_date_time
 FLAGS=-g -Wall
 
-all: bayrate check basicrate #copyMembership syncDBs
+all: bayrate check #basicrate copyMembership syncDBs
```
If we are removing these, we should clear the comment.

michaelhiiva reviewed (changes requested):
We have one warning: 

```
g++ -g -Wall -c db.cpp -o db.o -I/usr/include/mysql -lm -lgsl -lgslcblas -lmysqlpp -lboost_date_time
db.cpp: In member function 'void databaseConnection::syncDBs(std::map<int, tdListEntry>&, bool)':
db.cpp:712:7: warning: variable 'show' set but not used [-Wunused-but-set-variable]
  bool show = false;
       ^~~~
```

which we may want to address.

annabunches commented:
This PR ended up *way* out of scope from its original goal. In the future it would probably be a good idea to isolate our changes a little better. I came to add a dockerfile and now entire source files have been removed and that absolutely raises all of my danger instincts.

annabunches committed:
Remove comments referencing deleted code.

michaelhiiva commented:
There seem to be about three main applications issues here: Docker support, application instructions and removal of files for Dockerization. This could have been solved by not building them with docker in the first place; then, another pull request like you mentioned. But, I think removing those files here makes it clear not to put them into our docker build.

In hindsight, an issue for me is: you should have been able to handle the pull request without having commits by other members unless you had abandoned being assigned to this or had specifically requested this help or there were very small changes. My apologies for this.
